Fix crash when printing RGENGC_DEBUG=5 output from GC

KJTsanaktsidis opened the pull request:
I was trying to debug an (unrelated) issue in the GC, and wanted to turn
on the trace-level GC output by compiling it with -DRGENGC_DEBUG=5.
Unfortunately, this actually causes a crash in newobj_init() because the
code there tries to log the obj_info() of the newly created object.
However, the object is not actually sufficiently set up for some of the
things that obj_info() tries to do:

* The instance variable table for a class is not yet initialized, and
  when using variable-length RVALUES, said ivar table is embedded in
  as-yet unitialized memory after the struct RValue. Attempting to read
  this, as obj_info() does, causes a crash.
* T_DATA variables need to dereference their ->type field to print out
  the underlying C type name, which is not set up until newobj_fill() is
  called.

To fix this, we do two things:

* Firstly, we define a new function obj_info_basic(), which is just
  obj_info() except skipping bits that don't work on newly-created
  objects.
* Then, move the log message out of newobj_init() and into
  newobj_fill(); all possible codepaths which call newobj_init() do
  subsequently call newobj_fill(), so this won't change any output, and
  it allows the type of T_DATA objects to still be printed even in
  obj_info_basic(), which is helpful.

Bug: https://bugs.ruby-lang.org/issues/18795

nobu commented on gc.c:
```diff
@@ -13566,8 +13572,10 @@ str_len_no_raise(VALUE str)
 
-const char *
-rb_raw_obj_info(char *buff, const int buff_size, VALUE obj)
+#define OBJ_INFO_EXT 1 << 0
```
I think it is better to be parenthesized.

```suggestion
#define OBJ_INFO_EXT (1 << 0)
```

Or,

```suggestion
static const int OBJ_INFO_EXT = 1 << 0;
```

KJTsanaktsidis replied:
oh yeah, macros definitely need to be parenthesized, my bad. I'll change it to use the `static const int` version, because macros are the wrong kind of fun :)

nobu commented on gc.c:
```diff
@@ -13626,189 +13634,191 @@ rb_raw_obj_info(char *buff, const int buff_size, VALUE obj)
+        default:
+            break;
+        }
+    }
```
Indents seem broken.

KJTsanaktsidis replied:
I indented this `block` by one extra because I wrapped it in `if (obj_info_flags & OBJ_INOF_EXT)`. I can keep the indenting as-is if you prefer? Then it'll look like:

```
if (obj_info_flags & OBJ_INFO_EXT) {
switch(type) {
case T_NODE:
    ... etc
}
}
```

i.e. the if and the switch it contains will be indented the same. Is that what you want me to do?

nobu replied:
Seems inconsistent inside single block at least.
```C
        case T_NODE:
            UNEXPECTED_NODE(rb_raw_obj_info);
            break;
        case T_ARRAY:
                if (FL_TEST(obj, ELTS_SHARED)) {
                    // snip
                }
        // snip
            case T_SYMBOL: {
```
and
```C
        case T_IMEMO: {
                APPENDF((BUFF_ARGS, "<%s> ", rb_imemo_name(imemo_type(obj))));

            switch (imemo_type(obj)) {
            case imemo_ment:
                    {
                        // snip
                    }
            case imemo_iseq: {
            const rb_iseq_t *iseq = (const rb_iseq_t *)obj;
                    rb_raw_iseq_info(BUFF_ARGS, iseq);
            break;
            }
                case imemo_callinfo:
                    {
```

nobu commented:
Maybe you could simplify more by https://github.com/ruby/ruby/pull/6092?

KJTsanaktsidis commented:
@nobu yeah I'll have a look at rebasing this onto #6092 - looks like you've already split up parts of obj_info there, so I'll work out what parts can work on brand-new objects and adjust the call from `newobj_fill` accordingly

KJTsanaktsidis commented:
@nobu I rebased it - your changes in #6092 make this much neater now, thanks. What do you think?

jeremyevans commented:
@KJTsanaktsidis Sorry that we dropped the ball on this one.  Could you rebase against master (can drop the mjit specific changes) and then we can get this merged?

KJTsanaktsidis commented:
hah i'd completely forgotten about this too. I'll have another crack at rebasing this maybe this week.

KJTsanaktsidis commented:
@jeremyevans OK, i've rebased and re-tested this - the problem still exists without this patch, and the patch still fixes it.

I also changed my implementation to just copy-paste the three or four relevant lines into `obj_info_basic` and `rb_raw_obj_info_basic`. These functions are only a few lines anyway so I thought it didn't seem worth spraying a flags argument around the whole codebase just for this very unimportant edge case.
